[Federation] Move annotations and related parsing code as common code

irfanurrehman opened the pull request:
This PR moves some code, which was duplicate, around as common code.
Changes the names of structures used for annotations to common names.
s/FederatedReplicaSetPreferences/ReplicaAllocationPreferences/
s/ClusterReplicaSetPreferences/PerClusterPreferences/
This can be reused in job controller and hpa controller code.

**Special notes for your reviewer**:
@kubernetes/sig-federation-misc 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```NONE
```

Commit:
[Federation] Move annotations and related parsing code as common code

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44525)
<!-- Reviewable:end -->

irfanurrehman commented:
@k8s-bot verify test this
@k8s-bot cvm gce e2e test this

nikhiljindal commented:
Reassigning to @mwielgus since he wrote this code

irfanurrehman commented:
/assign @marun

marun commented:
Reviewed 14 of 14 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[federation/apis/federation/types.go, line 136 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikXlZPVQeyw8Dl6G2Y:-KikXlZPVQeyw8Dl6G2Z:bxbfnp2) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/apis/federation/types.go#L136)):*
> ```Go
> 	Rebalance bool
> 
> 	// A mapping between cluster names and preferences regarding local workload object (dep, rs, .. ) in
> ```

local -> a local ?

---

*[federation/apis/federation/types.go, line 147 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikYxVijjWDsPZl126_:-KikYxVijjWDsPZl126a:bxkovuh) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/apis/federation/types.go#L147)):*
> ```Go
> // a federated workload object.
> type PerClusterPreferences struct {
> 	// Minimum number of replicas that should be assigned to this Local Object. 0 by default.
> ```

Local Object -> cluster workload object ?

---

*[federation/apis/federation/types.go, line 151 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikZ-Kdbcr2qp0X_COb:-KikZ-Kdbcr2qp0X_COc:b-7ropln) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/apis/federation/types.go#L151)):*
> ```Go
> 	MinReplicas int64
> 
> 	// Maximum number of replicas that should be assigned to this Local Object. Unbounded if no value provided (default).
> ```

ditto

---

*[federation/apis/federation/types.go, line 155 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikZ1-dR6tjWX9wHgFa:-KikZ1-eAH0O-dFYOwAn:b-7ropln) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/apis/federation/types.go#L155)):*
> ```Go
> 	MaxReplicas *int64
> 
> 	// A number expressing the preference to put an additional replica to this Local Object. 0 by default.
> ```

ditto

---

*[federation/pkg/federation-controller/util/annotations.go, line 30 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikbCbSyR6W1CEZjcbB:-KikbCbSyR6W1CEZjcbC:b-90d5ck) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations.go#L30)):*
> ```Go
> // UnmarshalPreferences reads the preferences from the annotations on the given object.
> // It takes in an object and determines the supported types.
> // Callers need to pass the string key used to store the annotations.
> ```

This seems like YAGNI.  Why does the key need to vary?

---

*[federation/pkg/federation-controller/util/annotations.go, line 32 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-Kikb0ClqRckDiKPlaeD:-Kikb0ClqRckDiKPlaeE:b-ob86gc) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations.go#L32)):*
> ```Go
> // Callers need to pass the string key used to store the annotations.
> // Returns nil if the annotations with the given key are not found.
> func UnmarshalPreferences(obj runtime.Object, key string) (*fed.ReplicaAllocationPreferences, error) {
> ```

The name of this name seems vague considering the return type is specifically 'allocation preferences'.

---

*[federation/pkg/federation-controller/util/annotations.go, line 39 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikZXBY0Uw90bDphJPO:-KikZXBY0Uw90bDphJPP:b-noke7t) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations.go#L39)):*
> ```Go
> 	}
> 
> 	switch obj := obj.(type) {
> ```

Since annotations are stored in ObjectMeta, they can be accessed in a type-agnostic way:

```golang
// import meta "k8s.io/apimachinery/pkg/api/meta"

accessor, err := meta.Accessor(obj)
annotations := accessor.GetAnnotations()
```

Reference: https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/federation-controller/util/finalizers/finalizers.go#L28

---

*[federation/pkg/federation-controller/util/annotations_test.go, line 33 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikaXSRrxQyR84jOdCU:-KikaXSRrxQyR84jOdCV:bvl3odu) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations_test.go#L33)):*
> ```Go
> 
> const (
> 	TestPreferencesAnnotationKey = "federation.kubernetes.io/test-preferences"
> ```

Why not use the regular annotations key?

---

*[federation/pkg/federation-controller/util/annotations_test.go, line 36 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikaTpSrUIkaIH1_zTA:-KikaTpSrUIkaIH1_zTB:blrpbja) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations_test.go#L36)):*
> ```Go
> )
> 
> func TestParsePreferences(t *testing.T) {
> ```

This should probably match the name of the function being tested.


Also, why not use a test table and golang subtest?  This test is unnecessarily complicated otherwise.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
Review status: 7 of 14 files reviewed at latest revision, 10 unresolved discussions.

---

*[a discussion](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KioUi16u-nDrrALsQnm:-KioUi17AqKzDNLhXPIc:b-ksom1k) (no related file):*
Thanks @marun for the time. Have addressed your comments. PTAL!

---

*[federation/apis/federation/types.go, line 136 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikXlZPVQeyw8Dl6G2Y:-Kio1xdkYdxWvE7SIDKU:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/apis/federation/types.go#L136)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

local -> a local ?
</blockquote></details>

Done.

---

*[federation/apis/federation/types.go, line 147 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikYxVijjWDsPZl126_:-Kio1zPxx_bJaUknKIZB:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/apis/federation/types.go#L147)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Local Object -> cluster workload object ?
</blockquote></details>

Done.

---

*[federation/apis/federation/types.go, line 151 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikZ-Kdbcr2qp0X_COb:-Kio2-MXl3buYUHyd9AC:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/apis/federation/types.go#L151)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[federation/apis/federation/types.go, line 155 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikZ1-dR6tjWX9wHgFa:-Kio205GnO57wefVPJCx:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/apis/federation/types.go#L155)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[federation/pkg/federation-controller/util/annotations.go, line 30 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikbCbSyR6W1CEZjcbB:-Kio269jkdBhH2FZO85w:brmutwy) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations.go#L30)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

This seems like YAGNI.  Why does the key need to vary?
</blockquote></details>

The existing code used a different key for different objects. I agree that the key name need not vary on different objects, but I kept it as is to ensure not breaking existing usages. Also, I think the original idea was that different objects can have different annotation types (however using the same type as of now).
Please let me know your suggestions!

---

*[federation/pkg/federation-controller/util/annotations.go, line 32 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-Kikb0ClqRckDiKPlaeD:-Kio2g4Wk-ymOQm2_DCh:b4pobph) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations.go#L32)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

The name of this name seems vague considering the return type is specifically 'allocation preferences'.
</blockquote></details>

I have updated the name to GetAllocationPreferences. Please let me know if there are any better suggestions.. :)

---

*[federation/pkg/federation-controller/util/annotations.go, line 39 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikZXBY0Uw90bDphJPO:-Kio3LBsl9PleMJxvRjn:b-p6aovf) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations.go#L39)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Since annotations are stored in ObjectMeta, they can be accessed in a type-agnostic way:

```golang
// import meta "k8s.io/apimachinery/pkg/api/meta"

accessor, err := meta.Accessor(obj)
annotations := accessor.GetAnnotations()
```

Reference: https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/federation-controller/util/finalizers/finalizers.go#L28
</blockquote></details>

Thanks for the suggestion. My intention of type checking originally was to ensure an error on those federated object types which does not support preferences yet. But have updated it now, with a thought that we need not change this when new objects support the same preferences.

---

*[federation/pkg/federation-controller/util/annotations_test.go, line 33 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikaXSRrxQyR84jOdCU:-Kio4gd9VOJA-DLBPP3S:b9ac4c8) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations_test.go#L33)):*
> meta "k8s.io/apimachinery/pkg/api/meta"
I did this because as mentioned in another comment above, the existing keys used are not same on different objects.

---

*[federation/pkg/federation-controller/util/annotations_test.go, line 36 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikaTpSrUIkaIH1_zTA:-KioSniNcS0lRvrOnwvh:b-3ca1h5) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations_test.go#L36)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

This should probably match the name of the function being tested.


Also, why not use a test table and golang subtest?  This test is unnecessarily complicated otherwise.
</blockquote></details>

Updated the test accordingly!

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
@k8s-bot unit test this

irfanurrehman commented:
@k8s-bot unit test this

irfanurrehman commented:
@marun appreciate if you can u please relook, and check the fixes!!

marun commented:
Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[a discussion](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KioUi16u-nDrrALsQnm:-KjPRCJS9rCcMOrPCK50:b-366zfb) (no related file):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Thanks @marun for the time. Have addressed your comments. PTAL!
</blockquote></details>

/lgtm 

Sorry for the delay!

---

*[federation/pkg/federation-controller/util/annotations.go, line 30 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikbCbSyR6W1CEZjcbB:-KjPOOiRaQnCzH6bq3GG:b-ga1aqe) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations.go#L30)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

The existing code used a different key for different objects. I agree that the key name need not vary on different objects, but I kept it as is to ensure not breaking existing usages. Also, I think the original idea was that different objects can have different annotation types (however using the same type as of now).
Please let me know your suggestions!
</blockquote></details>

(No action required) I think it would be preferable to have a single annotation across types unless the semantics are going to differ.  I see now that this PR is maintaining the existing behavior, though, so unifying the annotation key can be tackled separately. 

---

*[federation/pkg/federation-controller/util/annotations.go, line 32 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-Kikb0ClqRckDiKPlaeD:-KjPOuyoQTYfziCIZDIs:brxp9sq) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations.go#L32)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

I have updated the name to GetAllocationPreferences. Please let me know if there are any better suggestions.. :)
</blockquote></details>

Sounds good.

---

*[federation/pkg/federation-controller/util/annotations.go, line 39 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikZXBY0Uw90bDphJPO:-KjPP--Wgf-9OmHteNGe:b-7i88ql) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations.go#L39)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Thanks for the suggestion. My intention of type checking originally was to ensure an error on those federated object types which does not support preferences yet. But have updated it now, with a thought that we need not change this when new objects support the same preferences.
</blockquote></details>

I later saw that you were using the meta accessor in testing, I apologize for assuming you didn't know about it.

---

*[federation/pkg/federation-controller/util/annotations.go, line 50 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KjPPQ9ZRoLlmUP-inQJ:-KjPPQ9ZRoLlmUP-inQK:b-awz734) ([raw file](https://github.com/kubernetes/kubernetes/blob/e976c561c15b11ad166a0c7d490570ddb8981889/federation/pkg/federation-controller/util/annotations.go#L50)):*
> ```Go
> 	}
> 
> 	var pref fed.ReplicaAllocationPreferences
> ```

Nice, I think variable locality aids in readability.

---

*[federation/pkg/federation-controller/util/annotations_test.go, line 33 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikaXSRrxQyR84jOdCU:-KjPPdTVDJuOpfgpN1i3:bm1snb0) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations_test.go#L33)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

> meta "k8s.io/apimachinery/pkg/api/meta"
I did this because as mentioned in another comment above, the existing keys used are not same on different objects.
</blockquote></details>

(No action required) While there's no harm in using a test-only key, I still don't see why it matters why key the test uses since GetAllocationPreferences is effectively type-agnostic.

---

*[federation/pkg/federation-controller/util/annotations_test.go, line 36 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikaTpSrUIkaIH1_zTA:-KjPR6LYXPDwNCfjof63:bmruvde) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations_test.go#L36)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Updated the test accordingly!
</blockquote></details>

Nice cleanup, thank you!

---

*[federation/pkg/federation-controller/util/annotations_test.go, line 36 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KjPPp47Prf__P_eczZC:-KjPPp47Prf__P_eczZD:bmb5pml) ([raw file](https://github.com/kubernetes/kubernetes/blob/e976c561c15b11ad166a0c7d490570ddb8981889/federation/pkg/federation-controller/util/annotations_test.go#L36)):*
> ```Go
> func TestGetAllocationPreferences(t *testing.T) {
> 	testCases := []struct {
> 		testname      string
> ```

(No action required) I think using a map instead of a slice is a slightly nicer way of providing the test name, e.g.

```golang
testCases := map[string]struct {
  prefs string
  ...
}{
  "good preferences": {
    prefs: {...}
   },
   ...
}  
for testName, testCase := range testCases {
...
}

``` 

---

*[federation/pkg/federation-controller/util/annotations_test.go, line 86 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KjPQaNEB0hF6TFS1l6D:-KjPQaNEB0hF6TFS1l6E:bchody3) ([raw file](https://github.com/kubernetes/kubernetes/blob/e976c561c15b11ad166a0c7d490570ddb8981889/federation/pkg/federation-controller/util/annotations_test.go#L86)):*
> ```Go
> 		pref, err := GetAllocationPreferences(tc.obj, TestPreferencesAnnotationKey)
> 		if tc.errorExpected {
> 			assert.Nil(t, pref)
> ```

(No action required) No harm, but I'm not sure there's a point in validating the state of prefs in the error case.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
Review status: 13 of 14 files reviewed at latest revision, 9 unresolved discussions.

---

*[a discussion](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KioUi16u-nDrrALsQnm:-KjS6KJxDXAwMqsR1a1K:b-7qzx8t) (no related file):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

/lgtm 

Sorry for the delay!
</blockquote></details>

No problems, Thanks for your time!

---

*[federation/pkg/federation-controller/util/annotations_test.go, line 33 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KikaXSRrxQyR84jOdCU:-KjS6YII6tFVE9Jpq3Hd:b-nnjupg) ([raw file](https://github.com/kubernetes/kubernetes/blob/11cf763bbe9f4d0b069c07402f97508ca92ea60e/federation/pkg/federation-controller/util/annotations_test.go#L33)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) While there's no harm in using a test-only key, I still don't see why it matters why key the test uses since GetAllocationPreferences is effectively type-agnostic.
</blockquote></details>

My point was also exactly that. It ideally does not matter which key is used by test. If different controllers define their own key name, and new controllers probably will use names not defined until now, why not just use a new name.

---

*[federation/pkg/federation-controller/util/annotations_test.go, line 36 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KjPPp47Prf__P_eczZC:-KjS7Cycd5r65RgI7tWE:byf6ndg) ([raw file](https://github.com/kubernetes/kubernetes/blob/e976c561c15b11ad166a0c7d490570ddb8981889/federation/pkg/federation-controller/util/annotations_test.go#L36)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) I think using a map instead of a slice is a slightly nicer way of providing the test name, e.g.

```golang
testCases := map[string]struct {
  prefs string
  ...
}{
  "good preferences": {
    prefs: {...}
   },
   ...
}  
for testName, testCase := range testCases {
...
}

``` 
</blockquote></details>

I like your suggestion, but skipping the update for now. Will positively use a similar structure in further PRs.
I should also say, I appreciate the learning. Please don't stop educating (me). :-)

---

*[federation/pkg/federation-controller/util/annotations_test.go, line 86 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-KjPQaNEB0hF6TFS1l6D:-KjS89ej1kt3eRMMsNTK:b-2znko7) ([raw file](https://github.com/kubernetes/kubernetes/blob/e976c561c15b11ad166a0c7d490570ddb8981889/federation/pkg/federation-controller/util/annotations_test.go#L86)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) No harm, but I'm not sure there's a point in validating the state of prefs in the error case.
</blockquote></details>

Updated this!

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
@marun Can u lgtm again, please.
@nikhiljindal @madhusudancs @quinton-hoole @csbell, Can any of you lgtm+approve please.

marun commented:
/lgtm

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44525#-:-KjZXTaelR8McXguhszr:b-8ks5qa)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
/assign @nikhiljindal @csbell

irfanurrehman commented:
/unassign @mwielgus

nikhiljindal commented on federation/apis/federation/types.go:
```diff
@@ -122,36 +122,40 @@ type ClusterList struct {
-// mentioned FederatedReplicaSet.
-type FederatedReplicaSetPreferences struct {
+// A set of preferences that can be added to federated version of workloads (deployments, replicasets, ..)
+// as a json-serialized annotation. The preferences allow the user to express in which clusters he/she
```
he/she -> they

irfanurrehman replied:
Done.

nikhiljindal commented on federation/apis/federation/types.go:
```diff
@@ -122,36 +122,40 @@ type ClusterList struct {
+// as a json-serialized annotation. The preferences allow the user to express in which clusters he/she
+// wants to put his/her replicas within the mentioned workload objects.
+type ReplicaAllocationPreferences struct {
 	// If set to true then already scheduled and running replicas may be moved to other clusters to
```
remove "to" at the end

irfanurrehman replied:
Done.

nikhiljindal commented on federation/apis/federation/types.go:
```diff
@@ -122,36 +122,40 @@ type ClusterList struct {
+type ReplicaAllocationPreferences struct {
 	// If set to true then already scheduled and running replicas may be moved to other clusters to
-	// in order to bring cluster replicasets towards a desired state. Otherwise, if set to false,
+	// in order to bring cluster replica workloads towards a desired state. Otherwise, if set to false,
```
change to "in order to match current state to the specified preferences"

irfanurrehman replied:
Done.

nikhiljindal commented on federation/apis/federation/types.go:
```diff
@@ -122,36 +122,40 @@ type ClusterList struct {
+	// If omitted, clusters without explicit preferences should not have any replicas scheduled.
 	// +optional
-	Clusters map[string]ClusterReplicaSetPreferences
+	Clusters map[string]PerClusterPreferences
```
optional: ClusterPreferences seems better than PerClusterPreferences to me.

irfanurrehman replied:
kept as is.

irfanurrehman replied:
Updated after second thoughts.. :)

nikhiljindal commented on federation/apis/federation/types.go:
```diff
@@ -122,36 +122,40 @@ type ClusterList struct {
+// A set of preferences that can be added to federated version of workloads (deployments, replicasets, ..)
+// as a json-serialized annotation. The preferences allow the user to express in which clusters he/she
+// wants to put his/her replicas within the mentioned workload objects.
+type ReplicaAllocationPreferences struct {
```
optional: How about calling this ReplicasDistributionPreferences

irfanurrehman replied:
Kept as is.

csbell commented:
/lgtm

nikhiljindal commented on federation/pkg/federation-controller/util/annotations.go:
```diff
@@ -0,0 +1,55 @@
+limitations under the License.
+*/
+
+package util
```
We want to have focused packages rather than a generic util package as a place to dump all util functions.
How about creating a `replicaspreferences` package at `federation/pkg/federation-controller/util/replicaspreferences`?

irfanurrehman replied:
Done.

nikhiljindal commented:
/lgtm cancel

marun commented:
@k8s-bot kops aws e2e test this

irfanurrehman commented:
@nikhiljindal, updated as per your review. PTAL!

irfanurrehman commented:
> @nikhiljindal, updated as per your review. PTAL!

@nikhiljindal waiting for an approve.

nikhiljindal commented:
Thanks!

/lgtm
/approve

madhusudancs commented:
/approve 
(Only for `hack/`. I am assuming @nikhiljindal has reviewed and approved the code in other directories)
